### PR TITLE
Prove the Milestone 0 contracts compose, not merely exist (KBR-32 / T-W9)

### PR DESCRIPTION
**Pull request by shelpuk**
Plan task **T-W9** — `.system_design/TEST_SUITE_IMPLEMENTATION_PLAN.md` §3 ·
design `.system_design/TEST_SUITE.md` §6.3.1, §7.2.1, §7.5.4 and the new §7.6 ·
Jira **[KBR-32](https://shelpuk.atlassian.net/browse/KBR-32)**.

Depends on T-W1 (KBR-24), T-W2 (KBR-25), T-W4 (KBR-27), T-W8 (KBR-31) — all merged.
No `src/kitty/` change: this adds evidence, not behaviour.

## Context for the Product Owner

Kitty Bridge sits between a coding agent and whichever LLM provider is being paid
for, and its core promise is that the conversation reaches the provider unchanged
apart from the edits it is licensed to make. Milestone 0 built three pieces of
machinery to check that promise: a fake provider that records exactly what
arrives, a way to stand a real bridge in front of it, and a shared vocabulary the
two describe requests in. Each landed working — **on its own**.

Nobody had yet run a single request through all three at once. That is this
change. It is a small ticket with disproportionate leverage: six later workstreams
are already planned on the assumption these pieces fit together, and until now
that assumption was untested. It now either holds or fails loudly on every commit.

It also closes three specific blind spots the earlier tickets could not reach,
because none of them started a real bridge:

- **One request in means one request out.** The bridge has an automatic retry
  system that fires when a provider returns nothing useful. It is invisible from
  the outside — the user still gets an answer — but it means paying for the same
  request several times. There was no test that would notice it firing when it
  should not. There is now.
- **A guard that nothing was watching.** One internal safety flag decides whether
  a streamed reply counts as real content. It is unreachable by ordinary testing;
  only a running bridge exercises it. It is now covered.
- **A slow failure is now a loud one.** That retry system costs 80 seconds of
  real waiting. A regression that switched it back on would have made the build
  slower rather than red — which has happened here before, costing roughly four
  minutes per pull request until it was found. The test now fails instead.

## What is in the diff

| File | Change |
|---|---|
| `tests/harness/test_vertical_slice.py` | New. Ten cases, ~0.5 s, `l1` by path default |
| `.system_design/TEST_SUITE.md` | New §7.6; §8.2's module registry and count updated |

## Measured before asserted

Every claim here was probed against a running bridge first, and two measurements
changed the design:

| Observation | Result |
|---|---|
| Healthy slice, non-streaming and streaming | 200, **0.01–0.02 s**, **1** capture |
| Default transport binding | Capture's query is **`""`** |
| Empty upstream reply, ladder delays flattened | **5** captures, 0.008 s |
| `has_content` forced false on a stream | The request **never completes** |
| `post()` timing out at 2 s against a sleeping ladder | Block exits after **62 s** |

- **The default binding carries no query string.** The falsification the plan
  names for this task — "a recorder that drops the query string fails it" — would
  have passed **vacuously**. The slice now drives a binding that carries one
  through `provider_config["base_url"]`, the product's own channel (KBR-143's
  merge rule), so the claim is about the bridge and not about the harness.
- **`post()`'s client timeout does not bound the slice.** Teardown waits for
  in-flight upstream handlers rather than aborting them. The wall-clock bound is
  therefore an assertion, not a reliance on that timeout.

## Three things the design review corrected

1. **The streaming case asserted a status that cannot fail.**
   `_stream_chat_completions` commits the downstream 200 with `sr.prepare()`
   before it opens the upstream, so every outcome on that route is a 200. The
   assertion is gone; the capture count is the whole claim there. The
   non-streaming case keeps its status assertion, where an upstream 400 is
   reachable and T-W8's `_RefusingTransport` is the defect that proves it bites.
2. **The query literal had no duplicate parameter name**, so half of what it
   claimed to prove was untestable. It is now `kbr32=slice%20value&dup=1&dup=2`.
3. **The budget's justification was wrong.** It claimed to catch "any ladder that
   sleeps". Any ladder that fires leaves an extra capture, which the count catches
   first and deterministically — and there are **two** ladders:
   `_EMPTY_RETRY_DELAYS` is read only by `_request_with_retry_single`, while every
   streaming path sleeps 1/2/4/8 s, so a streaming ladder fires twice inside a
   4-second budget. §7.6 now states the three mechanisms and the band each alone
   covers.

## The falsification sweep, and what it found

Plan §1.4 asks every harness to ship a deliberate defect it must detect. Both of
this module's falsification cases drive the **shipped** assertions — named
helpers — rather than copies of them, because a copied assertion falsifies
nothing about the one that ships.

Each of the recorder's capture fields was then mutated in turn and the module
re-run. Ten of eleven mutations were caught by exactly one case each. One was not:

> **A recorder reading `request.path` instead of `rel_url.raw_path` killed
> nothing** — the adapter's own path is `/v1/messages`, which contains nothing
> percent-encoded, so the two are identical on this route and §7.2.1's trap was
> untestable. The base URL now carries an encoded path prefix as well, and the
> mutation fails the field case.

That is the second claim in this task found vacuous because the *driven request*
did not carry the thing the claim was about, and it is the argument for running
the sweep rather than reasoning about it. §7.6 records the result, including the
one assertion that survives on purpose (and why) and one that is weaker than it
looks.

## Cost and safety

- Module runtime **~0.5 s** (0.42–0.54 s over three runs); §8.2 updated with it.
- The ladder falsification flattens the delays with `monkeypatch` rather than
  waiting them out — **values flattened, lengths unchanged**, since
  `max_attempts` is computed from the list lengths. An 80-second falsification
  case is the defect commit `691e974` fixed once already.
- The transports defined here are **never registered**. Registration is an
  import-time side effect on module-global state, so a stray `register_transport`
  would leave this module green alone and the full suite red under a different
  collection order. `test_bridge.py`'s registry meta-test is green.
- All measurements are Python 3.12.3; CI across 3.10–3.13 is the authority for
  the budget. A version-dependent failure means raising it, never skipping.

## Verification

- `pytest tests/harness/ -q --strict-markers` — **706 passed**
- `ruff check .` — clean · `mypy src/kitty` — clean · `lint-imports` — 5/5 kept
- `ruff format` applied to the new file only
- Layer guards (`test_layer_markers.py`, `test_layer_selection.py`) — 68 passed

## Out of scope, stated so it is not read as an omission

Corpus-wide quantification is T-D1's (KBR-51); the other five wire formats and
the three custom transports are T-B1–T-B3's; non-loopback addressing stays with
T-E1/T-E2. Bounding fixture *teardown* against a fired ladder is real — the 62
seconds above — but aborting in-flight handlers would change T-W8's module and is
a decision §7.3 already took for the proxy; it is left to T-K6 and T-B4.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018b37S3LGuQABeEDr8jeeLv

[KBR-32]: https://shelpuk.atlassian.net/browse/KBR-32?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/harness/test_vertical_slice.py [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+            "id": "msg_empty",
+            "type": "message",
+            "role": "assistant",
+            "model": "recorder-model",
+            "content": [],
+            "stop_reason": "end_turn",
+        }
+    ).encode()
+    await reply.begin(200, {"content-type": "application/json", "content-length": str(len(payload))})
+    await reply.write(payload)
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_vertical_slice.py:353]: Fixed in c1f41db — accepted, and it was the right catch.

The failure mode named is the one a falsification case must never have: if aiohttp stopped ending a prepared response implicitly, this case would **hang** rather than go red, and a hang in the gate reads as "slow" rather than "broken". Relying on that was not a deliberate trade-off, just an omission.

`_empty_reply` now matches `RecordingUpstream._default_responder`'s non-streaming reply exactly: content length through the attribute, then `begin`, `write`, `write_eof`. Re-measured after the change — still five captures with the ladder delays flattened, so nothing behavioural moved.

(Also replying on the duplicate of this comment from the earlier round, id 3996089281.)

File: tests/harness/test_vertical_slice.py [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_vertical_slice.py:353]: Duplicate of id 3996103091, which the same review raised on the newer commit. Fixed in c1f41db — `_empty_reply` now calls `write_eof()` explicitly and matches `_default_responder`'s shape. See the reply there for the reasoning.

[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_vertical_slice.py:353]: Partly accepted, but the premise here is not right, so recording the correction rather than the change alone.

`_default_responder` does **not** write its non-streaming reply chunked. It sets `response.content_length = len(payload)` at `tests/harness/recorder.py:654` and then writes — chunked encoding is that responder's *streaming* branch (`Content-Type: text/event-stream`), which is a different path. `_empty_reply` is non-streaming, so declaring a length was already correct and already matched the shipped response shape.

What was genuinely inconsistent was the **form**: I set `content-length` as a literal header while the sibling sets the `content_length` attribute. That is now aligned in c1f41db, along with the explicit `write_eof()` from id 3996103091, so the two helpers read identically.

File: tests/harness/test_vertical_slice.py [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+        budget of zero is over-budget for any real request, so what this proves
+        is that :func:`_drive` — the one path every case in this module takes —
+        actually applies it.
+        """
+        with pytest.raises(AssertionError) as excinfo:
+            await _drive(AiohttpTransport(WireFormat.ANTHROPIC_MESSAGES), budget=0.0)
+
+        message = str(excinfo.value)
+        assert "budget" in message
+        assert str(_EMPTY_LADDER_SECONDS) in message
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_vertical_slice.py:704]: Accepted — asserted both in c1f41db.

The question was worth asking rather than assuming, and the answer follows from what the message is for. Those two numbers are the whole diagnostic: they are what lets a reader tell a fired retry ladder from a merely slow CI runner without a rerun. Asserting one of the pair left the other removable with nothing going red, which is the same shape this module objects to elsewhere.

File: tests/harness/test_vertical_slice.py [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+        # Sampling keys are closed (§3.3.1b), so only the ones this body can
+        # carry are lifted; anything else would raise rather than pass silently.
+        sampling = {key: body[key] for key in ("max_tokens", "temperature") if key in body}
+
+        return Request(
+            envelope=Envelope(model=body.get("model"), stream=body.get("stream")),
+            conversation=Conversation(turns=turns, sampling=sampling),
+            consumed=frozenset({"model", "messages", "stream", *sampling}),
+            source=body,
+        )
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_vertical_slice.py:321]: Accepted as a scope comment, in c1f41db — a comment, not a test.

The assumption is real: `MessagesReader` lifts `m["content"]` straight into `Text(...)`, and Anthropic Messages also permits a list of content blocks. Everything this module drives comes from `minimal_inbound_body`, which sends a plain string, so it holds today.

Not pinning it with a test, deliberately. `MessagesReader` is evidence that a capture is readable **at all** by something shaped like a `Projection` — it is not a projection of the format, and §7.4.1 closes that set at six `reader_<format>.py` modules. `harness/reader_anthropic_messages.py` (T-A1) owns the content-array shape and tests it. Growing this helper toward generality would quietly make it a seventh reader, which is the thing §7.5.4 and §7.4.1 between them are trying to prevent — so the comment says what it does not read, and why that is someone else's job.

File: .system_design/TEST_SUITE.md [hunk 5]
```diff
@@ -2938,11 +3111,13 @@ business, together with the job that runs them; doing it earlier would remove th
...
-**Six modules are bulleted below, and `tests/cli/test_stream_encoding.py` (KBR-10) is described
-after them — seven in all, named here so T-K6 inherits a list rather than a search** — the count
+**Seven modules are bulleted below, and `tests/cli/test_stream_encoding.py` (KBR-10) is described
+after them — eight in all, named here so T-K6 inherits a list rather than a search** — the count
 is what T-K6 and T-H1 plan against. (The bullet count and the KBR-10 paragraph were already
 drifting apart before T-W8 added two; spelling out both is what stops the next addition
-guessing which set it joins.)
+guessing which set it joins. T-W9 joins the **bulleted** set, not the paragraph above it, which
+still names `tests/test_egress_https_proxy.py` and `tests/harness/test_connect_proxy.py`
+separately.)
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply .system_design/TEST_SUITE.md:3577]: Clarified in c1f41db, though the count itself was already right — worth setting out, since the comment reads the list as seven bullets.

It is **five** bullets naming **seven** modules, because the T-W4 and T-W8 rows name two modules each:

| Bullet | Modules |
|---|---|
| KBR-132 | `test_tls_certs.py` |
| T-W4 | `test_recorder.py`, `test_recorder_falsification.py` |
| T-W8 | `test_bridge.py`, `test_bridge_falsification.py` |
| T-W9 | `test_vertical_slice.py` |
| KBR-144 | `test_responses_string_input.py` |

1 + 2 + 2 + 1 + 1 = seven bulleted, plus the KBR-10 paragraph = eight. §8.2 counts **modules**, which is the unit T-K6 and T-H1 plan against — bullets would be the wrong unit for that.

So this was readability, not arithmetic, and the ambiguity predates this PR (§8.2 already carries a parenthetical saying the two sets were drifting). Since I touched that sentence anyway, the lead now says "in five bullets, since the T-W4 and T-W8 rows name two modules each", which removes the miscount this comment demonstrates is possible. Splitting KBR-10 into its own bullet is a larger edit to a section this PR is only registering itself in, so leaving that.

File: tests/harness/test_vertical_slice.py [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+
+**Layer.** No ``pytestmark``, so these take the ``l1`` path default, following
+T-W4 and T-W8. The reason is §8.2's and only §8.2's: ``l3`` is in
+``PENDING_ACTIVATION_LAYERS``, so an ``l3`` marker today would leave the slice
+checked by no job at all. §8.2 names this module so T-K6 inherits a list.
+
+**Evidence base.** Every number quoted here was measured on Python 3.12.3. CI
+across 3.10–3.13 is the authority; a version-dependent timing failure means
+raising :data:`_SLICE_BUDGET_SECONDS`, never skipping a case.
+"""
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_vertical_slice.py:71]: Not taking this one — the mapping already exists, closer to the code than a front-matter table would put it.

Each test class sits under a banner naming its requirements: `# -- R1, R5 — one request in, one request out`, `# -- R2 — the capture satisfies T-W2's declared contract`, and so on through R7 (lines 453, 507, 571, 601, 621, 645). A reader cross-referencing a claim finds it beside the class rather than by counting rows in a table hundreds of lines above.

Adding a test-class column would duplicate that mapping in a second place, and a duplicated mapping is one that can go stale: renaming a class would leave the table silently wrong, with nothing to catch it. The banners cannot drift the same way, because they are adjacent to what they label.

The table's own job is narrower and I would rather keep it there — for each claim, *why only a driven bridge can show it*, which is the argument for this ticket existing at all.

**Comment by shelpuk** [comment]
Review loop closed — **round 3 returned "No findings"**, and all 12 checks pass on `c1f41db`, including the four Python legs (3.10 / 3.11 / 3.12 / 3.13, ~19:30–19:56 each). That matters for one claim in particular: the 4.0 s wall-clock budget was the single thing the pre-merge review named as able to flip its approval on a slower runner, and it held on every version.

### What the three rounds changed

| Round | Raised | Outcome |
|---|---|---|
| 1 (`e124556`) | 1 warning, 5 suggestions | 3 fixed, 1 corrected, 1 declined |
| 2 (`7d3f521`) | The round-1 warning, still standing | Fixed in `c1f41db` |
| 3 (`c1f41db`) | — | No findings |

The warning was the one worth having. `_empty_reply` relied on aiohttp ending a prepared response implicitly, where `RecordingUpstream._default_responder` finishes its own explicitly. The failure mode is specific and bad: if that implicit behaviour ever changed, this **falsification** case would hang rather than go red — and a hang in an ~18.5-minute gate reads as "slow", not "broken". A case whose whole job is to fail loudly must not be able to hang. Fixed, and re-measured at five captures, unchanged.

### Two corrections to the record

Small, but this review is the durable artefact, so:

1. **I did push back on one finding, and correct another** — the round-3 summary says there were "no rounds where the author pushed back". The docstring-table-ordering suggestion is declined with reasoning ([thread](https://github.com/Shelpuk-AI-Technology-Consulting/kitty-bridge/pull/80#discussion_r3996089292)): the requirement-to-class mapping already exists as banner comments adjacent to each class, and duplicating it in a front-matter table creates something that goes stale silently on a rename. Separately, the Content-Length finding's premise was wrong — `_default_responder` does set a content length on its non-streaming reply (`recorder.py:654`); chunked is its *streaming* branch ([thread](https://github.com/Shelpuk-AI-Technology-Consulting/kitty-bridge/pull/80#discussion_r3996089291)). The form was aligned anyway.

2. **The budget is measured around the driven request, not fixture entry-to-exit.** `time.monotonic()` brackets the `post()` call inside the fixture block. The review's conclusion is right — it deliberately does not bound `BridgeFixture.stop`, which §7.5 says cannot be bounded — but the window is narrower than described, and that is the point: teardown waiting out in-flight handlers (measured at 62 s) is exactly the noise the assertion must not absorb.

### Standing caveat, unchanged

The bound makes a reintroduced retry ladder **red**; it does not make it **cheap**. A red case still pays up to ~72 s of teardown. That is recorded in §7.6 and §8.2 for T-K6 and T-B4 rather than fixed here, because aborting in-flight handlers would change T-W8's module and is a decision §7.3 already took separately for the proxy.

Not merging — leaving that to you.

**Comment by shelpuk** [comment]
Windows leg triaged. **Four failures, one of them mine — fixed in `c1f8857`. The other three are pre-existing and belong to other tickets.**

macOS and all four Linux legs passed. Every failure traces to one root cause: **Windows' clock cannot resolve two adjacent events**, the same property §8.3 already carries exemptions for (KBR-188).

### Mine — fixed

`test_the_budget_is_enforced_on_every_driven_request` → `DID NOT RAISE`.

It passed `budget=0.0` on the reasoning that zero is over-budget for any real request. That is false on Windows: `elapsed` measures **exactly 0.0**, so `0.0 <= 0.0` holds and the assertion never fires. Reproduced locally by freezing `time.monotonic` at `1402.328` — the literal value the Windows leg reported:

```
budget= 0.0 -> DID NOT RAISE          <-- the Windows failure
budget=-1.0 -> raised AssertionError  <-- the fix
```

The impossible budget is now **negative**. An elapsed time cannot fall below zero at any clock resolution, so the case is deterministic rather than a bet on the runner's granularity. The 4.0 s budget is unaffected — an elapsed of 0.0 sits comfortably inside it, and the module's nine other cases passed on Windows.

### Not mine — three, and they need their own tickets

| Failure | Owner | Why |
|---|---|---|
| `test_provider_aiohttp.py::…[check_arrival_increases]` — `arrival must strictly increase … got [1402.328, 1402.328]` | **T-B1 / KBR-40** | A raw failure with **no exemption row**. PR #83 merged at 14:01, and the Windows leg (#79) landed after it — so this module has **never run on Windows**. It needs the exemption row T-W4 already has. |
| `test_recorder.py::…[check_arrival_increases]` | **KBR-188** | `UnexpectedExemptionPass` — the exempt assertion *passed*. |
| `test_recorder_falsification.py::…miscount_only_connection_check` | **KBR-188** | Same. |

The last two are the more interesting finding, and worth stating plainly: **§8.3 models these as "always false on Windows", but the underlying behaviour is a race.** Whether two adjacent requests land in the same clock tick is chance. PR #79 saw them fail (so the exemptions were satisfied, leg green); this run saw them pass (so `UnexpectedExemptionPass` fired, leg red). Both directions, same assertions, different runs.

That makes the Windows leg **nondeterministic today, independent of this PR** — and it will bite every branch, not just this one. An exemption can express "false here"; it cannot express "usually false here", which is what this actually is.

I have not touched any of the three. Two are a registry only KBR-188 should change, and the third is a missing row that belongs in KBR-40's ticket rather than smuggled into this one — `ci-required` will stay red until they are addressed, so flagging rather than absorbing.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by shelpuk** [comment]
Re-run result — **my fix worked, and the picture narrowed sharply.**

**Windows: 1 failed, 5158 passed** (previous run: 4 failed).

- `test_vertical_slice.py` **does not appear in the failures at all**. The negative budget in `c1f8857` holds on Windows.
- **Both KBR-188 `UnexpectedExemptionPass` failures vanished without any change to them.** Nobody touched `tests/exemptions.py`, `test_recorder.py` or `test_recorder_falsification.py` between the two runs. That is direct evidence for the diagnosis in my previous comment: those assertions are a **race**, not a deterministic platform truth. Two runs, same code, opposite outcomes.

### What remains — one failure, and it is not this PR's

```
FAILED tests/harness/test_provider_aiohttp.py::TestTheRecorderPassesEveryConformanceCheck
       ::test_per_session_check[check_arrival_increases]
AssertionError: arrival must strictly increase in sent order; got [1402.328, 1402.328].
```

**T-B1 / KBR-40.** Timeline, which is the whole explanation:

| When | What |
|---|---|
| 14:01 | PR #83 merges `tests/harness/provider_aiohttp.py` |
| later | PR #79 adds the Windows leg |

So this module has **never executed on Windows**. `RecordingUpstream` already carries `recorder-arrival-increases-per-session` for exactly this assertion; the provider recorder needs the equivalent row and has none.

Unlike the KBR-188 pair, this one is not a coin flip in the same way — with no exemption at all, it fails whenever the clock is coarse enough, which on that runner is often.

### Deliberately not fixed here

Per the repo's own rule that a change should trace to its ticket, and §8.3's that the registry is acknowledged debt with an owner: adding a row for **KBR-40**'s module inside **KBR-32**'s PR would put another ticket's platform debt under this one's name, and §8.3 is a registry whose whole value is that each row names the ticket that owes it.

**Two things need their own tickets**, and I could not file them — the Jira integration dropped mid-session:

1. **KBR-40** — `tests/harness/provider_aiohttp.py` needs the `check_arrival_increases` exemption row for Windows. One row, the sanctioned mechanism, blocking `ci-required` on every branch until it lands.
2. **KBR-188** — §8.3 models these assertions as *"false on Windows"*; they are *"nondeterministically false on Windows"*. An exemption can express the first and not the second, so today the leg fails in **both** directions: when the assertion fails without a row, and when it passes with one (`UnexpectedExemptionPass`). The two runs on this PR demonstrate both. This is a design gap in the mechanism, not a missing row.

Nothing in this PR's own diff is red. Everything else is green: all four Linux legs, macOS, CodeQL, and the automated review.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by shelpuk** [comment]
Correction to my earlier triage, now that Jira is reachable again: **the remaining Windows failure is already filed, already diagnosed, and already fixed in an open PR.** I should not have proposed filing anything new.

### The remaining failure is [KBR-208](https://shelpuk.atlassian.net/browse/KBR-208), fixed in [#87](https://github.com/Shelpuk-AI-Technology-Consulting/kitty-bridge/pull/87)

`tests/harness/recorder.py` stamps `arrival` with `time.monotonic()`, which on **Windows + Python ≤ 3.12** is backed by `GetTickCount64` at **15.625 ms** resolution. Two loopback requests inside one tick get the same timestamp. `perf_counter()` in `c1645d8` fixes it.

The ticket's arithmetic check reproduces on this PR's own failure: `1402.328 = 89749 × 1/64 s`, an exact multiple of the tick. Same fingerprint, different run.

I filed nothing new. I added the second sighting to KBR-208 and the mechanism finding to [KBR-188](https://shelpuk.atlassian.net/browse/KBR-188).

### One finding that is new, and it is actionable on #87

**`c1645d8` fixes the cause but leaves five exemption rows the fix makes obsolete** — `tests/exemptions.py` is untouched by that PR. Five of its six rows are keyed on the coarse clock, all KBR-188's; the sixth is KBR-189's abort race and is unaffected. Each says the assertion is *false on Windows*; on `perf_counter()` it becomes *true*, so `ratchet` raises `UnexpectedExemptionPass` — *"delete the row"* — for all five at once, and the leg goes red from the other side.

**This PR already ran that failure**, which is what makes it a prediction rather than a guess. Two Windows runs, identical code, nothing changed in between:

| Run | Result |
|---|---|
| 1 | 4 failed — including `UnexpectedExemptionPass` on two KBR-188 rows, i.e. the arrival assertions **passed** |
| 2 | 1 failed — only the raw `check_arrival_increases` |

On a 15.6 ms tick that outcome is occasional. On `perf_counter()` it is every run, for all five rows. Raised on #87 with the suggestion to delete them in the same commit.

### Where that leaves this PR

Nothing in this diff is red. The one failing test is KBR-208's, in a module this PR does not touch, and it clears when #87 merges — at which point this branch needs a `main` merge and nothing else.

My own Windows failure was real and is fixed in `c1f8857`: the impossible time budget was `0.0`, and on that same coarse clock a fast request measures **exactly** `0.0`, so `0.0 <= 0.0` held and the guard never fired. It is negative now — unsatisfiable at any clock resolution. Same root cause as KBR-208, reached from the other end.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[KBR-208]: https://shelpuk.atlassian.net/browse/KBR-208?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[KBR-188]: https://shelpuk.atlassian.net/browse/KBR-188?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

File: tests/harness/test_vertical_slice.py [hunk 7]
```diff
@@ -0,0 +1,782 @@
...
+
+    # The capture count is reported first and is the diagnosis, not decoration:
+    # the gate runs ~18.5 minutes on contended runners, so "slow" must be
+    # distinguishable from "a ladder fired" without a rerun.
+    diagnosis = (
+        "exactly one upstream request, so this is a slow runner and not a retry "
+        "ladder — raise the budget rather than hunting a regression"
+        if len(captures) == 1
+        else f"{len(captures)} upstream requests, so a retry ladder fired"
+    )
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/harness/test_vertical_slice.py]: Accepted — fixed in `3e5253a`. This is a correctness bug in a diagnostic, and the reasoning is exactly right.

A retry ladder **cannot** leave zero captures. Zero means the bridge reached some other upstream — §7.5.4's decoy — and `_assert_one_upstream_request` says so in words three lines away, which makes the two-arm message contradict this module's own vocabulary as well as itself.

The reachability argument is what makes it worth fixing rather than noting. The budget only fires in the **4–10 s band** (`post()`'s `DEFAULT_TIMEOUT` outranks it above that), and a decoy pointed at a *slow* wrong upstream lands squarely in it with nothing captured. §7.6 justifies this message by saying a slow runner and a fired ladder must be distinguishable without a rerun — in that arm it made them distinguishable in the **wrong direction**, which is worse than saying nothing.

Three arms now. Verified by driving both counts and reading the message back rather than assuming:

```
ZERO captures  -> zero upstream requests, so the bridge reached some other upstream
                  entirely and no retry ladder could have fired
ONE capture    -> exactly one upstream request, so this is a slow runner and not a
                  retry ladder — raise the budget rather than hunting a regression
```

File: .system_design/TEST_SUITE.md [hunk 8]
```diff
@@ -3120,6 +3120,202 @@ saying so here stops a future reader reading it as an unpaid debt.
...
+for the number, across every leg — 3.10–3.13 on Linux plus the pinned Windows and macOS legs
+§8.4 added after this module was written — and a platform- or version-dependent failure means
+raising it, never skipping.
+
+**The slice needs no Windows exemption row, and must not acquire one.** §8.3's registry is all
+arrival *ordering*: Windows' clock cannot separate two adjacent requests, so every "arrival
+increases" assertion is false there (KBR-188). This module asserts only that `arrival` is
+**populated and typed**, which a coarse clock satisfies — ordering is T-W4's claim, not the
+slice's. Tightening it into an ordering check here would add debt to §8.3 for a claim that is
+already made, and made better, one layer down.
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply .system_design/TEST_SUITE.md:3332]: Accepted — added in `3e5253a`. A completeness point, and the right one: §8.2 points T-K6 at that paragraph for this module's platform story, so an incomplete account there is a trap rather than an omission.

The paragraph recorded only the **first** way the coarse clock reached this module (`arrival` asserted as typed rather than ordered, which was foreseen). The **second** was found by the Windows leg itself: the enforcement case compared `elapsed` against a literal `0.0`, which on Windows measures exactly `0.0`, so `0.0 <= 0.0` held and the assertion could not fire — green on five legs, "DID NOT RAISE" on the sixth.

Your point about the **remedy** is the part most worth having in the document, and I have written it as its own paragraph rather than a sentence. Those two paragraphs now say different things on purpose:

- *"a platform-dependent failure means raising the budget, never skipping"* governs the **4.0 s** margin — a judgement about runner speed.
- This was not that. It was an assertion written so that it **could not fire** on a coarse clock, and the fix made it resolution-independent rather than widening anything.

Conflating them would teach the next author to reach for a bigger number when the defect is that the comparison is unfalsifiable. And as you note, the paragraph's existing claim still holds: no §8.3 row was added, because the platform dependence was **removed** rather than amnestied.
